torchtext/experimental/vocab.py

```diff
@@ -0,0 +1,177 @@
+from collections import OrderedDict
+import logging
+from typing import List
+
+import torch
+import torch.nn as nn
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+
+def _infer_shape(f):
+    num_lines = 0
+    for line in f:
+        num_lines += 1
+    f.seek(0)
+    return num_lines
+
+
+def vocab_from_file_object(file_like_object, **kwargs):
+    r"""Create a `Vocab` object from a csv file like object.
+
+    Note that the tensor corresponding to each vector is of type `torch.float`.
+
+    Format for csv file:
+        token1
+        token2
+        ...
+        token_n
+
+    Args:
+        file_like_object (FileObject): a file like object to read data from.
+        Remaining keyword arguments: Passed to the constructor of Vocab class.
+
+    Returns:
+        Vocab: a `Vocab` object.
+    """
+    ordered_dict = OrderedDict()
+    num_lines = _infer_shape(file_like_object)
+    for line in tqdm(file_like_object, unit_scale=0, unit="lines", total=num_lines):
+        token = line.rstrip()
+        if token in ordered_dict:
+            ordered_dict[token] += 1
+        else:
+            ordered_dict[token] = 1
+
+    return Vocab(ordered_dict, **kwargs)
+
+
+class Vocab(nn.Module):
+    r"""Creates a vocab object which maps tokens to indices.
+
+    Arguments:
+        ordered_dict (collections.OrderedDict): object holding the frequencies of each token found in the data.
+        min_freq: The minimum frequency needed to include a token in the vocabulary.
+            Values less than 1 will be set to 1. Default: 1.
+        specials: The tuple of special tokens (e.g., padding or eos) that will be prepended/postpended to the vocabulary.
+            based on the `specials_first` flag. The ordering of the tuple will be preserved. Default: [<pad>']
+        specials_first: Whether to add special tokens into the vocabulary at first. If it is False,
+            they are added into the vocabulary at last. Default: True.
+
+    Raises:
+        ValueError: if a default `unk_token` isn't provided.
+    """
+
+    def __init__(self, ordered_dict, min_freq=1, unk_token='<unk>', specials=('<pad>',), specials_first=True):
+        super(Vocab, self).__init__()
+
+        if not unk_token:
+            raise ValueError("A default unk token wasn't provided.")
+
+        tokens = []
+        for token, freq in ordered_dict.items():
+            if freq >= min_freq:
+                tokens.append(token)
+
+        # assume unk_token and special tokens dont appear in ordered_dict
+        if specials_first:
+            tokens = [unk_token] + list(specials) + tokens
```

Comment:
`specials` tokens don't include `unk_token`, which is in counterintuitive. Here we are assuming that '<unk>' is always indexed by "0". What's the solution for the following case:
a "pre-trained" vocab from a list of tokens - ['<pad>', '<sep>', '<unk>', 'this', 'is', 'pytorch']

Reply:
This is a good point. After our offline discussion, it seems like the best approach is to ensure that the `unk_token` is in the `specials` list and that the name of the `unk_token` used is provided.